src/layout.rs

```diff
@@ -288,171 +277,142 @@ fn split(area: Rect, layout: &Layout) -> Rc<[Rect]> {
-                });
-
-                match *size {
-                    Constraint::Min(v) | Constraint::Max(v) => {
-                        ccs.push(elements[i].width | EQ(WEAK) | f64::from(v));
-                    }
-                    _ => {}
-                }
-            }
-        }
-        Direction::Vertical => {
-            for pair in elements.windows(2) {
-                ccs.push((pair[0].y + pair[0].height) | EQ(REQUIRED) | pair[1].y);
-            }
-            for (i, size) in layout.constraints.iter().enumerate() {
-                ccs.push(elements[i].x | EQ(REQUIRED) | f64::from(dest_area.x));
-                ccs.push(elements[i].width | EQ(REQUIRED) | f64::from(dest_area.width));
-                ccs.push(match *size {
-                    Constraint::Length(v) => elements[i].height | EQ(MEDIUM) | f64::from(v),
-                    Constraint::Percentage(v) => {
-                        elements[i].height | EQ(MEDIUM) | (f64::from(v * dest_area.height) / 100.0)
-                    }
-                    Constraint::Ratio(n, d) => {
-                        elements[i].height
-                            | EQ(MEDIUM)
-                            | (f64::from(dest_area.height) * f64::from(n) / f64::from(d))
-                    }
-                    Constraint::Min(v) => elements[i].height | GE(MEDIUM) | f64::from(v),
-                    Constraint::Max(v) => elements[i].height | LE(MEDIUM) | f64::from(v),
-                });
-
-                match *size {
-                    Constraint::Min(v) | Constraint::Max(v) => {
-                        ccs.push(elements[i].height | EQ(WEAK) | f64::from(v));
-                    }
-                    _ => {}
-                }
-            }
-        }
-    }
-    solver.add_constraints(&ccs).unwrap();
-    for &(var, value) in solver.fetch_changes() {
-        let (index, attr) = vars[&var];
-        let value = if value.is_sign_negative() {
-            0
-        } else {
-            value as u16
-        };
-        match attr {
-            0 => {
-                results[index].x = value;
-            }
-            1 => {
-                results[index].y = value;
-            }
-            2 => {
-                results[index].width = value;
-            }
-            3 => {
-                results[index].height = value;
-            }
-            _ => {}
-        }
-    }
+    {
+        use taffy::prelude::*;
+        let mut taffy = Taffy::with_capacity(layout.constraints.len());
+        let mut taffy_keys = Vec::with_capacity(layout.constraints.len());
+
+        // disabling rounding, because otherwise the following tests fail:
+        // widgets_table_columns_widths_can_use_mixed_constraints
+        // widgets_table_column_spacing_can_be_changed
+        //
+        // but disabling rounding makes the following tests fail:
+        // layout::tests::test_vertical_split_by_height
+        taffy.disable_rounding();
```

Comment:
this needs to be discussed, opening a "thread"

Reply:
For these tests, I'd generally approach this by adding simpler unit tests that cover the same functionality on the current implementation first to work out exactly which calculation is failing.

That assert in `test_vertical_split_by_height` should just assert the actual Rects instead of properties of the rects - doing so would give the information necessary to debug what's different. It's probably a demonstration of one of things I believe is actually a bug not a feature - 10 rows split into [10%, max(5), min(1)] should lead to [1,5,4], not [1,5,3] (or whatever the actual result is).

Reply:
> That assert in test_vertical_split_by_height should just assert the actual Rects instead of properties of the rects - doing so would give the information necessary to debug what's different.

i know *why* `test_vertical_split_by_height` is failing - it is because rounding is disabled (read the comment again), because `10% of 9` is `<1.0`, which because rounding is disabled, gets trimmed to `0`

i had disabled rounding because of the other 2 mentioned tests, and i wanted feedback on that first before deciding on a path, because it seems to me that `widgets_table_columns_widths_can_use_mixed_constraints` at least is now actually outputting the correct things (no extra gap at the end?)
as for why `widgets_table_column_spacing_can_be_changed` is failing, i currently have no clue, but it is likely related to the previous test failing

---

also i have not encountered tests that test specific lengths that something should give and so some thing may be missed, example `test_vertical_split_by_height` currently tests 2 things: 1. that all elements add up to the size specified, and 2. that the next elements are always `y <= y`, no specific tests on which the size the elements themself should be

Reply:
>also i have not encountered tests that test specific lengths that something should give 

Yeah, that's exactly the sort of thing I see as a problem (a problem I've spent many days dealing on in legacy software maintenance in $PREVIOUS_JOB). testing a second order thing rather than a first order (e.g. asserting collection.is_empty() = true rather than checking collection = []) makes broken tests annoying to fix.

Reply:
> Oh - and I think probably keeping rounding enabled seems like the intuitively right choice

i now had a look at the original cassowary code, it does *not* use rounding, it just trims of the decimal point:
https://github.com/ratatui-org/ratatui/blob/37fa6abe9d5dc459dc9855ea10f06afa72717c98/src/layout.rs#L387
(context: `value` is `f64`)



<details>
<summary>some example value from running <code>examples/layout.rs</code> with some extra logging where the width x height is 116x57:</summary>

```txt
[2023-08-08T12:40:46.727+02:00 WARN  ratatui::layout]: res_f64 [
    ResultingRects {
        x: 0.0,
        y: 0.0,
        width: 116.0,
        height: 4.0,
    },
    ResultingRects {
        x: 0.0,
        y: 4.0,
        width: 116.0,
        height: 28.5,
    },
    ResultingRects {
        x: 0.0,
        y: 32.5,
        width: 116.0,
        height: 24.5,
    },
]
[2023-08-08T12:40:46.728+02:00 WARN  ratatui::layout]: result [
    Rect {
        x: 0,
        y: 0,
        width: 116,
        height: 4,
    },
    Rect {
        x: 0,
        y: 4,
        width: 116,
        height: 28,
    },
    Rect {
        x: 0,
        y: 32,
        width: 116,
        height: 24,
    },
]
```

</details>

<details>
<summary>The same test run against current PR (rounding disabled, using trimming), with similar logging</details>

```txt
[2023-08-08T12:45:56.635+02:00 WARN  ratatui::layout]: taffy f32 [
    ResultingRects {
        x: 0.0,
        y: 0.0,
        width: 116.0,
        height: 4.0,
    },
    ResultingRects {
        x: 0.0,
        y: 4.0,
        width: 116.0,
        height: 17.458824,
    },
    ResultingRects {
        x: 0.0,
        y: 21.458824,
        width: 116.0,
        height: 35.541176,
    },
]
[2023-08-08T12:45:56.635+02:00 WARN  ratatui::layout]: results [
    Rect {
        x: 0,
        y: 0,
        width: 116,
        height: 4,
    },
    Rect {
        x: 0,
        y: 4,
        width: 116,
        height: 17,
    },
    Rect {
        x: 0,
        y: 21,
        width: 116,
        height: 35,
    },
]
```

</details>